File: src/Coachable/1-DSA/4-String,Sorting/MergeSort.py
"""
MergeSort.py
"""

"""
Merge Sort

Stable/unstable sorting
Time: O(n.log n)
Space: O(n) -> Extra space for auxiliary array
"""


def merge_sort(array: list[int]) -> None:
    aux_array = [0 for _ in range(len(array))]  # auxiliary array
    divide(array, 0, len(array) - 1, aux_array)


def divide(array: list[int], left, right, aux_array):
    # base condition
    if left >= right:
        return

    mid = left + (right - left) // 2
    divide(array, left, mid, aux_array)
    divide(array, mid + 1, right, aux_array)

    merge(array, left, mid, right, aux_array)


def merge(array, low, mid, high, aux_array):
    # copy items to aux-array
    for idx, num in enumerate(array):
        aux_array[idx] = num

    i, j = low, mid + 1

    for k in range(low, high + 1):
        if i > mid:
            array[k] = aux_array[j]
            j += 1
        elif j > high:
            array[k] = aux_array[i]
            i += 1
        elif is_less(aux_array[i], aux_array[j]):
            array[k] = aux_array[i]
            i += 1
        else:
            array[k] = aux_array[j]
            j += 1


def is_less(item1, item2):
    return item1 < item2


# arr = [8, -5, 10, 2, 1]
# arr = [-2, 3, -5]
arr = [3, 8, 2, 15, 27, 21, 17, 10, 16, 7, 24, 0, 4, 6, 18, 5]
merge_sort(arr)
print(arr)

File: src/misc/Learning Material/sorting/MergeSort.java
package Coachable.sorting;

public class MergeSort {

	private MergeSort() {
	}

	private static Comparable<Object>[] aux;


	public static void sort(Comparable[] arr) {

		// initiate the auxiliary array
		aux = new Comparable[arr.length];

		sort(arr, 0, arr.length - 1);
	}

	private static void sort(Comparable<Object>[] arr, int low, int high) {
		// base condition for recursive calls
		// merge sort needs at-least 2 elements
		if (low >= high) return;

		int mid = low + ((high - low) / 2);

		sort(arr, low, mid);
		sort(arr, mid + 1, high);

		merge(arr, low, mid, high);
	}

	private static void merge(Comparable<Object>[] arr, int low, int mid, int high) {

		if (high - low + 1 < 1) return;

		System.arraycopy(arr, 0, aux, 0, arr.length);

		// two pointers. i: low to mid, j:mid+1 to high
		// k: low to high
		int i = low, j = mid + 1;
		for (int k = low; k <= high; k++) {
			if (i > mid) arr[k] = aux[j++];
			else if (j > high) arr[k] = aux[i++];
			else if (isLess(aux[i], aux[j])) arr[k] = aux[i++];
			else arr[k] = aux[j++];
		}
	}

	private static boolean isLess(Comparable<Object> a, Comparable<Object> b) {
		if (a.equals(b)) return false;
		return a.compareTo(b) < 0;
	}

}
